File: 2020/08_handheld_halting/aoc08.py
"""Handheld Halting

Advent of Code 2020, day 8
Solution by Geir Arne Hjelle, 2020-12-08
"""
import pathlib
import sys
import parse

debug = print if "--debug" in sys.argv else lambda *_: None

INSTRUCTION = parse.compile("{operation} {argument:d}")


def acc(argument):
    """Change accumulator

    acc increases or decreases a single global value called the accumulator by
    the value given in the argument. For example, acc +7 would increase the
    accumulator by 7. The accumulator starts at 0. After an acc instruction,
    the instruction immediately below it is executed next.
    """
    return 1, argument


def nop(argument):
    """Do nothing

    nop stands for No OPeration - it does nothing. The instruction immediately
    below it is executed next.
    """
    return 1, 0


def jmp(argument):
    """Jump to a new instruction

    jmp jumps to a new instruction relative to itself. The next instruction to
    execute is found using the argument as an offset from the jmp instruction;
    for example, jmp +2 would skip the next instruction, jmp +1 would continue
    to the instruction immediately below it, and jmp -20 would cause the
    instruction 20 lines above to be executed next.
    """
    return argument, 0


INSTRUCTIONS = {"acc": acc, "nop": nop, "jmp": jmp}


def run_code(code):
    """Run the given code until a loop is detected"""
    pointer, accumulator = 0, 0
    visited = [False] * len(code) + [True]

    while not visited[pointer]:
        visited[pointer] = True
        instruction = INSTRUCTION.parse(code[pointer])

        dptr, dacc = INSTRUCTIONS[instruction["operation"]](instruction["argument"])
        pointer += dptr
        accumulator += dacc

    return accumulator, pointer == len(code)


def create_mutant(code, line_num):
    """Create a mutant by changing the given line num"""
    instruction = INSTRUCTION.parse(code[line_num])
    if instruction["operation"] == "acc":
        return code, False

    mutant = code.copy()
    op = "nop" if instruction["operation"] == "jmp" else "jmp"
    mutant[line_num] = f"{op} {instruction['argument']}"
    debug(f"{line_num:3d} {code[line_num]} -> {mutant[line_num]}")

    return mutant, True


def main(args):
    """Solve the problem for all file paths"""
    for file_path in [pathlib.Path(p) for p in args if not p.startswith("-")]:
        solve(file_path)


def solve(file_path):
    """Solve the problem for one file path"""
    print(f"\n{file_path}:")
    code = file_path.read_text().strip().split("\n")

    # Part 1
    accumulator, _ = run_code(code)
    print(f"The original code accumulates {accumulator}")

    # Part 2
    for line_num in range(len(code)):
        mutant, changed = create_mutant(code, line_num)
        if not changed:
            continue

        accumulator, ended = run_code(mutant)
        if ended:
            print(f"The fixed code accumulates {accumulator}")
            break


if __name__ == "__main__":
    main(sys.argv[1:])

File: 2020/08_handheld_halting/aoc08.jl
# Encoding Error
#
# Advent of Code 2020, day 8
# Solution by Geir Arne Hjelle, 2020-12-09

module AOC08

using Pipe


#
# Instructions
#

# Accumulate the given value
function op_acc(argument; mutated=false)
    ReturnValue(1, argument)
end

# Move to the next instruction
function op_nop(argument; mutated=false)
    ReturnValue(if mutated argument else 1 end, 0)
end

# Jump relative
function op_jmp(argument; mutated=false)
    ReturnValue(if mutated 1 else argument end, 0)
end

INSTRUCTIONS = Dict("acc" => op_acc, "nop" => op_nop, "jmp" => op_jmp)

struct Instruction
    operation
    argument::Int32

    function Instruction(line)
        op, arg = split(line)
        new(INSTRUCTIONS[op], parse(Int32, arg))
    end

    function Instruction(op, arg)
        new(op, arg)
    end
end

struct ReturnValue
    d_pointer::Int32
    d_accumulator::Int32
end


#
# Run programs
#

# Execute one instruction
function execute(instruction, pointer, accumulator; mutated=false)
    ret_val = instruction.operation(instruction.argument, mutated=mutated)
    pointer + ret_val.d_pointer, accumulator + ret_val.d_accumulator
end


# Run the full program until an infinite loop is detected
function run(program; mutated_line=-1)
    pointer, accumulator = 1, 0
    visited = vcat(zeros(Bool, length(program)), true)

    while !visited[pointer]
        visited[pointer] = true
        pointer, accumulator = execute(
            program[pointer], pointer, accumulator, mutated=pointer == mutated_line
        )
    end

    accumulator, pointer > length(program)
end


# Solve the problem for one file
function solve(input)
    # Parse input
    instructions = split(input, "\n") .|> Instruction

    # Part 1
    part_1 = instructions |> run |> first

    # Part 2
    part_2 = @pipe (
        1:length(instructions)
        .|> run(instructions, mutated_line=_)
        |> filter(r -> r[2], _)
        |> first
        |> first
    )

    part_1, part_2
end


# Solve the problem for one file
function solve_file(file_path)
    println("\n$(file_path)")
    input = open(file_path) do fid
        read(fid, String) |> strip
    end
    input .|> solve
end

# Solve the problem for each file
[a for a in ARGS if a[1] != '-'] .|> solve_file .|> s -> join(s, "\n") |> println

end  # module
